### DISCUSSION TITLE
Speech to text streaming WAV files

### DISCUSSION THREAD

**Discussion by Grzetan** [deepgram/community#677] [General help] [answered]
Hello,
I have a problem with streaming WAV files to deepgram's speech to text service.

I have a queue of small WAV files encoded as `base64` strings (about `1s` duration) that need to be send to deepgram for transcription in real time. The problem is that deepgram only processes the first WAV file sent, for all of the other ones transcription in empty. 

I suppose deepgram expects one big WAV file splitted into chunk, not a lot of small WAV files. I tried to join those small files but it doesn't work. If the first small chunk is sent without modifications and all of the others are sent without a header (only payload) the result is the same, empty transcription.

Should I modify payload length in the header to some big number so deepgram expects more payload to come in?
Or maybe I should convert my audio chunk to some other format?

I work in python and i'm using deepgram python SDK.

Here is some code from my app:
```python
def remove_first_x_bytes(base64_string, x):
    if len(base64_string) < x:
        return ""
    decoded_bytes = base64.b64decode(base64_string)
    sliced_bytes = decoded_bytes[x:]
    return base64.b64encode(sliced_bytes).decode("utf-8")

async def speech_to_text(session_queue: SessionQueues) -> None:
    """
    This function is ran as a never ending task
    """
    config = DeepgramClientOptions(options={"keepalive": "true"})
    client: DeepgramClient = DeepgramClient(DEEPGRAM_API_KEY, config)
    ws = client.listen.live.v("1")
    options = LiveOptions(
        model="nova-2",
        language="pl-PL",
        smart_format=True,
    )
    leave_header: bool = True

    def on_message(self, result, **kwargs):
        nonlocal session_queue
        sentence = result.channel.alternatives[0].transcript
        logger.info(sentence)
        if len(sentence) == 0:
            return

        session_queue.output_queue.put_nowait(SpeechToTextOutput(sentence).to_dict())

    def on_error(self, error: ErrorResponse, **kwargs):
        logger.info(error)

    def on_metadata(self, metadata, **kwargs):
        logger.info(metadata)

    ws.on(LiveTranscriptionEvents.Transcript, on_message)
    ws.on(LiveTranscriptionEvents.Error, on_error)
    ws.on(LiveTranscriptionEvents.Metadata, on_metadata)
    ws.start(options)

    while True:
        audio = await session_queue.input_user_audio_queue.get()

        if not leave_header:
            audio = remove_first_x_bytes(audio, 44)  # 44 == WAV header size
        leave_header = False  # After the first chunk, remove header

        ws.send(base64.b64decode(audio))
        await asyncio.sleep(0.1)
```

Thanks in advance for any help

**Comment by team-deepgram** [comment]
Thanks for asking your question about Deepgram! If you didn't already include it in your post, please be sure to add as much detail as possible so we can assist you efficiently, such as:
- The `request_id` if you have a question about your requests or transcription responses.
- The features you used or the full `api.deepgram.com` URL you sent your request to, including parameters.
- Any code snippets you can share.

**Comment by jkroll-deepgram** [comment]
Hi @Grzetan, I see you're also using our streaming transcription - it will likely be simpler to use pre-recorded transcription by submitting the files. Link to our [pre-recorded Python SDK starter](https://developers.deepgram.com/docs/getting-started-with-pre-recorded-audio#transcribe-a-local-file). Then you can send each file in its own API request and get the JSON transcription response for it.

**Reply by Grzetan** [reply]
I don't think it will help in my case. The audio files I get are coming from user's microphone and they are also really small (like I said 1s) so sending each chunk individually would be like transcribing only 1s of user speech.

**Reply by jkroll-deepgram** [reply]
@Grzetan are they actually consecutive audio chunks from a user speaking, or are you stitching them together? I would expect that you can stream continuous audio from a microphone to Deepgram. Here's a [streaming microphone code example](https://github.com/deepgram/deepgram-python-sdk/tree/main/examples/streaming/microphone) with the Python SDK if that helps. If that's not what you're getting at, it would be helpful if you can describe your use case and how you're obtaining these different brief audio snippets.

**Reply by prashnm** [reply]
@Grzetan HI, I have exactly similar issue. Can you please tell me if you are able to solve it. For me i want to send microphone data from front end to python deepgram speech to text api. I havent found a way to do that. 
if you did that please do let me know? 

@jkroll-deepgram

**Reply by jkroll-deepgram** [reply]
Hi @prashnm, we present a few tools for integrating a frontend with a Python backend using Deepgram in [Starting Out with Python and Deepgram Live Streaming Audio](https://deepgram.com/learn/python-deepgram-roundup).

**Reply by prashnm** [reply]
@jkroll-deepgram Hey Thanks a lot, I was looking for type of solution for days. This really solved most of my troubles.

I just have one more question,  is there a way that I can pass metadata and retrive it back along with transcript. i tried few ways, by passing using url and also here 
 self.deepgram_socket = await dg_client.transcription.live({'punctuate': True, 'interim_results': False}, metadata={tags:'user_id')
 
 in both methods i could not see the user_id being returned with data.

**Reply by jkroll-deepgram** [reply, marked as answer]
@prashnm We have an "extra" metadata field, that sounds like what you're looking for: [Extra Metadata](https://developers.deepgram.com/docs/extra-metadata).

**Reply by davidvonthenen** [reply]
hi @prashnm 

I would take a look at tagging and metadata fields:
https://developers.deepgram.com/docs/tagging

Depending on the SDK you are using, you can look at the options/schema objects for the respective SDKs.

Python live:
https://github.com/deepgram/deepgram-python-sdk/blob/main/deepgram/clients/live/v1/options.py#L93-L95

Python prerecorded:
https://github.com/deepgram/deepgram-python-sdk/blob/main/deepgram/clients/prerecorded/v1/options.py#L126-L128
